### DISCUSSION TITLE
Crazyflie doesn't take off with flow and ranger decks installed under assist or test script control

### DISCUSSION THREAD

**Discussion by jjdike** [bitcraze/discussions#1170] [Q&A]
Hi - my crazyflie seems to be working as expected with just the flow deck.  The hover assist mode and command based flight controls work.  When I add the ranger deck, the crazyflie connects and the flow and ranger decks are recognized based on the console output.  However the thrust doesn't seem to be enough to get the crazyflie off the ground with these assisted commands.  With the game controller and without flight assist I can lift off, but with the hover assist and command based flight control it won't lift off.  Same thing for the sample ranger deck python script - the 'push' example.  The crazyflie connects and turns on but the script just runs for a short time without enough thrust to lift off then terminates on its own.  This is with battery voltage close to 4 V.

**Comment by knmcguire** [comment]
Hi!

It seems more like an estimation issue then if you are able to take off with manual flight. 

If you look at the flight tab's position estimate, or plot the position estimate values, does they seem right to you?

**Comment by jjdike** [comment]
I reinstalled the deck - the long pin set seems just barely long enough for a flow deck below and the multiranger deck on top, capturing the battery.  This time I can hover from the cfclient or do command based flight from the cfclient without a controller attached.  However when I try the push script https://www.bitcraze.io/documentation/tutorials/getting-started-with-stem-ranging-bundle/ , the crazyflie takes off but then lands right away.  If I comment out the line about checking for multiranger.up 

if is_close(multiranger.UP):
    keep_flying = False

crazyflie will keep hovering but only reacts to something being placed below it, not to the sides or above.  If I print out the range value in the 'is_close' function, it first has values of 'none' then they all become 0.0.

One last piece of information is that if I check the parameters from cfclient and look at the list for 'decks', it looks like the flow and multiranger are found.  Is bcZRanger2 part of the flow deck?

Thanks.

![image](https://github.com/bitcraze/discussions/assets/157736336/7ff029ef-d55f-4f65-9714-22cb4c010311)

update - I took the multiranger deck off and put it back on and was back to the crazyflie no longer taking off.  I also saw the state estimates drifting.  I removed the multiranger deck and it flew normally again.  Not sure if the deck is just not being seated well or if something else is wrong.   

To see if perhaps there was an issue with the amount of pin engagement, I let the battery dangle so I could push the multiranger deck on further, then held the crazyflie while I connected with cfclient and tried a command based hover.  Only two of the motors were thrusting when held level.  The motors kept running even after I clicked on the 'land' button, and even after I clicked the 'disconnect' button from cfclient.
Repeating this test with the multiranger deck removed but everything else the same showed normal behavior for the motor thrust levels .

**Reply by knmcguire** [reply]
There seems to be something wrong with your multiranger then. If you look at the console tab in the cfclient, what kind of error messages do you see? Coul you copy past it here?

**Reply by jjdike** [reply]
I started cfclient with the crazyflie having both the flow deck installed on the bottom and the multiranger deck on top.  This is the console window output from cfclient.  This was after connecting to crazyflie, clicking the 'takeoff' button in the cfclient command based flight box, then clicking 'land'.  The crazyflie behavior was to turn on at 'takeoff', though underpowered so it wouldn't actually take off, then it kept going after I clicked 'land' and disconnected from crazyflie through cfclient.  I pushed the power button on crazyflie to turn it off.

SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:e527895e8c6e (2023.11) CLEAN
SYS: I am 0x33373336343247060018002E and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcMultiranger for deck 0
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2023.11
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
MR: Init front sensor [OK]
MR: Init back sensor [OK]
MR: Init up sensor [OK]
MR: Init left sensor [OK]
MR: Init right sensor [OK]
DECK_CORE: Deck 0 test [OK].
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 16296 bytes
STAB: Starting stabilizer loop
ESTKALMAN: State out of bounds, resetting
SUP: Can not fly
SUP: Ready to fly

**Reply by knmcguire** [reply]
hmmm this is very strange... so the decks don't indicate that there is anything wrong. I'll ask some more questions in a new thread.

**Comment by knmcguire** [comment]
Some additional questions to help out further:
- You are sure that the orientation is correct of the decks: https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#orientation. Also, the multiranger is all the way on the top and the flowdeck on the bottom?
- How about the battery? Have you done a battery test (button next to console tab).
I'll see if anybody else here internally at bitcraze has any ideas...

**Comment by jjdike** [comment]
I believe the orientation is correct - flow deck with sensor pointing downward underneath the main deck, battery on top of the main deck, then the multiranger deck with upward sensor facing up and the "up/front" markings visible from the top and aligned with the front of the crazyflie.

```
Here's the output from connecting to crazyflie with just the flow deck attached:
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:e527895e8c6e (2023.11) CLEAN
SYS: I am 0x33373336343247060018002E and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2023.11
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 17608 bytes
STAB: Starting stabilizer loop
ESTKALMAN: State out of bounds, resetting
SUP: Can not fly
SUP: Ready to fly
```

Here are the results of the battery test with just the flow deck attached from the console window:
```
HEALTH: Idle:3.81V sag: 0.57V (< 0.70V) [OK]
ESTKALMAN: WARNING: Kalman prediction rate off (96)
```

Crazyflie correctly responds to the command based flight buttons from cfclient with just the flow deck attached.

Here is console output after shutting crazyflie off, installing multiranger deck on top, reconnecting to crazyflie, then running the battery test and then the propeller test from the console window.

```

SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:e527895e8c6e (2023.11) CLEAN
SYS: I am 0x33373336343247060018002E and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2023.11
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 17608 bytes
STAB: Starting stabilizer loop
ESTKALMAN: State out of bounds, resetting
SUP: Can not fly
SUP: Ready to fly

HEALTH: Idle:3.79V sag: 0.49V (< 0.70V) [OK]
ESTKALMAN: WARNING: Kalman prediction rate off (96)
STAB: WARNING: stabilizer loop rate is off (950)


HEALTH: Acc noise floor variance X+Y:0.004876, (Z:24.936344)
HEALTH: Motor M1 variance X+Y: 2.83 (Z:10.30), voltage sag:0.36
HEALTH: Motor M2 variance X+Y: 0.70 (Z:4.80), voltage sag:0.34
HEALTH: Motor M3 variance X+Y: 0.61 (Z:0.35), voltage sag:0.38
HEALTH: Motor M4 variance X+Y: 1.38 (Z:6.57), voltage sag:0.37
HEALTH: Propeller test on M1 [FAIL]. low: 0.0, high: 2.50, measured: 2.83
ESTKALMAN: WARNING: Kalman prediction rate off (15)
```

I reran the propeller test a couple of more times and they all passed on those subsequent tests.

**Reply by knmcguire** [reply]
Sorry for the late reply! (I caught the flu this week).

It seems that for the second try, the console didn't detect the multiranger at all. You mentioned earlier that the pins were hardly large enough to properly connect the multiranger, and perhaps that is the core of the problem. Everytime the crazyflie takes off it disconnects with the multiranger so that is absolutely a problem. Are you using the long pins that came with the crazyflie or [the pin header pins that you get with the thicker battery](https://store.bitcraze.io/products/350mah-battery-including-pins)? There are even longer pins that you could use as well: https://store.bitcraze.io/products/male-long-deck-connector.

**Comment by jjdike** [comment]
Thanks for the reply.  Glad you are feeling better.

I am using the long pins that come with crazyflie and the extra long pins
that are available should arrive tomorrow.  I did try letting the battery
dangle while holding the crazyflie flies in some of the tests I mentioned
so I could push the multiranger deck further onto the pins but not getting
different results.  I also have another multiranger deck on the way so
hopefully these gets me the expected behavior from the flow amd multiranger
deck combo.

**Reply by knmcguire** [reply]
Ah oke! That would be the next thing that I would suggest. Please keep us up to date. If it is a hardware issue than we should probably replace the faulty multiranger

**Comment by jjdike** [comment]
I just installed the longer header pins and tried with my original
multiranger deck - same behavior as before.  It won't take off correctly
with the command based controls..  I replaced my original multiranger deck
with a new one and the command based controls work correctly and the 'push'
example script runs correctly - so it seems like there is something wrong
with the multiranger deck.

**Reply by knmcguire** [reply]
Jup! No doubt about that it is a hardware issue. Please send us an email at contact _at_ bitcraze.io with a link to this discussion and your order number and we will send a replacement.
